jaxopt/_src/levenberg_marquardt.py

```diff
@@ -326,70 +368,73 @@ def update(self, params, state: NamedTuple, *args, **kwargs) -> base.OptStep:
     # Current value of the loss function F=0.5*||f||^2.
     loss_curr = state.loss
 
+    # TODO: clean up the linear_solver to take matrix directly and then clean up
+    # linear equation solves below for materialize_jac=True to call them.
+
     # For geodesic acceleration, we calculate  jtrpp=JT * r",
     # where J is jacobian and r" is second order directional derivative.
     if self.materialize_jac:
-      damping_term = state.damping_factor * jnp.identity(params.size)
       # Note that instead of taking the inverse of jtj_corr and multiply that
-      # by state.gradient, we are using `solve`, which uses LU decomposition of
+      # by state.gradient, we prefer to use `jsp.solve`, which uses Cholesky of
       # jtj_corr and uses that to obtain velocity. This has the advantage of
       # lower number of floating point operations and therefore less numerical
       # error which can be helpful for the case of single precision arithmatics.
-      jtj_corr = state.jtj + damping_term
-      velocity = jnp.linalg.solve(jtj_corr, state.gradient)
+      if self.solver == 'qr':
+        damping_term = jnp.sqrt(state.damping_factor) * jnp.identity(params.size)
+        aug_jac = jnp.vstack([state.jac, damping_term])
+        aug_res = jnp.hstack([state.residual, jnp.zeros(params.size)])
+        q, r = jsp.linalg.qr(aug_jac, mode='economic')
+        velocity = jsp.linalg.solve_triangular(r, q.T @ aug_res)
+      else:
+        damping_term = state.damping_factor * jnp.identity(params.size)
+        jtj_corr = state.jtj + damping_term
+        if self.solver == 'lu':
```

Comment:
Thanks for all the work Amir! 
I'm really sorry but I'm still confused about the difference between the 'lu' and the 'inv' case in this if statement. The lines 392: `velocity = jnp.linalg.solve(jtj_corr, state.gradient)` and 394 `velocity = jnp.linalg.solve(jtj_corr, state.gradient)` are identical to me. If the difference lies in the way state.gradient or jtj_corr are defined, it would be better to add a note about it to ease the reading.

Reply:
Great! Thank you, I would not be surprised if the inv does not give very accurate solutions.